### DISCUSSION TITLE
Environments, views, and PYTHONPATH

### DISCUSSION THREAD

**Discussion by ser3n1ty87** [spack/spack#30657] [Q&A] [answered]
I have been playing around with spack and so far I am amazed. 
However, I have a question regarding the necessity of PYTHONPATH within the context of environments and views.

I completely understand the necessity of setting the PYTHONPATH for the "regular" case where you would load some python package with `spack load <my-package>` as it got installed into its own prefix and the python interpreter has to be informed about this location.

What I don't understand is why we would still need setting this PYTHONPATH when using enviroments and its views.
Using the default settings for the creation of an environment (including a view then), python packages are put into the `/<env-name>/.spack-env/view/lib/python3.9/site-packages` folder. Running `python3 -m site` tells me that python knows about this exact location which is not surprising as it's just its "own" site-packages folder. But still PYTHONPATH gets set in this case pointing to the aforementioned site-packages folder of the interpreter within view and also to the site-packages folders of the python packages itself `<prefix-of-python-package>/lib/python3.9/site-packages`, which seems redundant to me. Simply unsetting the PYTHONPATH also shows that (at least in my little test case) it is in fact redundant as importing the modules with python still works.

So my question is: Are there python packages which do things completely different and even in the context of views the PYTHONPATH is necessary or is it safe to just always unset the PYTHONPATH when using environments with activated views?
Is the set PYTHONPATH maybe just a misconfiguration of environments on my side?

The reason why I am trying to avoid a globally set PYTHONPATH is that it constantly leads to errors and confusion for me.
For example the python system interpreter (with a different version) would use packages from the PYTHONPATH and also virtual environments don't work as expected as the PYTHONPATH effectively breaks isolation.

**Comment by adamjstewart** [comment, marked as answer]
You are correct that PYTHONPATH is unnecessary with views (either inside an environment or not). I don't think Spack has any mechanisms for treating env vars differently based on whether you are using a view or not, so if it's set for individual packages, it also gets set for views. I think your best bet is to simply unset PYTHONPATH after activating the environment.
